chemprop/cli/hyperopt.py

```diff
@@ -1,16 +1,348 @@
+    )
+
+    checkpoint_config = CheckpointConfig(
+        num_to_keep=args.num_checkpoints_to_keep,
+        checkpoint_score_attribute="val_loss",
+        checkpoint_score_order=monitor_mode,
+    )
+
+    run_config = RunConfig(
+        checkpoint_config=checkpoint_config,
+    )
+
+    ray_trainer = TorchTrainer(
+        lambda config: train_model(config, args, train_loader, val_loader, logger),
+        scaling_config=scaling_config,
+        run_config=run_config,
+    )
+
+    tune_config = tune.TuneConfig(
+        metric="val_loss",
+        mode=monitor_mode,
+        num_samples=args.num_samples,
+        scheduler=scheduler,
+    )
+
+    tuner = tune.Tuner(
+        ray_trainer,
+        param_space={"train_loop_config": build_search_space(args.search_parameter_keywords, args.epochs)},
+        tune_config=tune_config,
+    )
+
+    return tuner.fit()
+
+def main(args: Namespace):
+
+    format_kwargs = dict(
+        no_header_row=args.no_header_row,
+        smiles_cols=args.smiles_columns,
+        rxn_cols=args.reaction_columns,
+        target_cols=args.target_columns,
+        ignore_cols=args.ignore_columns,
+        weight_col=args.weight_column,
+        bounded=args.loss_function is not None and "bounded" in args.loss_function,
+    )
+    featurization_kwargs = dict(
+        features_generators=args.features_generators, keep_h=args.keep_h, add_h=args.add_h
+    )
+
+    train_data, val_data, test_data = build_splits(args, format_kwargs, featurization_kwargs)
+    train_dset, val_dset, test_dset = build_datasets(args, train_data, val_data, test_data)
+
+    X_d_scaler, V_f_scaler, E_f_scaler, V_d_scaler = normalize_inputs(
+        train_dset, val_dset, args
+    )
+    input_scalers = {"X_d": X_d_scaler, "V_f": V_f_scaler, "E_f": E_f_scaler, "V_d": V_d_scaler}
+
+    if "regression" in args.task_type:
+        scaler = train_dset.normalize_targets()
+        val_dset.normalize_targets(scaler)
+        logger.info(f"Train data: mean = {scaler.mean_} | std = {scaler.scale_}")
+    else:
+        scaler = None
+
+    train_loader = MolGraphDataLoader(
+        train_dset, args.batch_size, args.num_workers, seed=args.data_seed
+    )
+    val_loader = MolGraphDataLoader(val_dset, args.batch_size, args.num_workers, shuffle=False)
+    if test_dset is not None:
+        test_loader = MolGraphDataLoader(
+            test_dset, args.batch_size, args.num_workers, shuffle=False
+        )
+    else:
+        test_loader = None
+
+    torch.manual_seed(args.pytorch_seed)
+
+    model = build_model(args, train_loader.dataset)
+    monitor_mode = "min" if model.metrics[0].minimize else "max"
+
+    results = tune_model(args, train_loader, val_loader, logger, monitor_mode)
+
+    best_result = results.get_best_result()
+    best_config = best_result.config
+    best_checkpoint = best_result.checkpoint  # Get best trial's best checkpoint
+    best_result_df = best_result.metrics_dataframe  # Get best result as pandas dataframe
```

Comment:
I think we should save the results as a CSV, but could we save results for all the trials? I think it would be helpful to have the comparison for all the different configurations saved for the user to see.

Reply:
By default, Tune logs results for TensorBoard, CSV, and JSON formats (https://docs.ray.io/en/latest/tune/tutorials/tune-output.html). It would be great to check whether it does as what we expect.

Reply:
I also have a question about whether the output table will contain information on values for metrics for each trail other than `val_loss` and the used hyperparameters. I guess probably yes as we already add those information to the log during `validation_step`. 
To be more specific, we have below two lines in the validation step ([Link](https://github.com/chemprop/chemprop/blob/v2/dev/chemprop/models/model.py#L159-L160)):

> 
        self.log_dict(metric2loss, batch_size=len(batch[0]))
        self.log("val_loss", losses[0], batch_size=len(batch[0]), prog_bar=True)

Reply:
@am2145 Thanks for the suggestion! I have added code to save the results for all trials.
@shihchengli I checked and Tune does log results in the TensorBoard format. I was able to view the trial results in TensorBoard. I also added the code for Tune to save the trial results in a CSV file. Here is a snippet of the results CSV file generate by ray tune. It saves all the information such as the used hyperparameters.


```
train_loss,val/mse,val_loss,epoch,step,timestamp,checkpoint_dir_name,should_checkpoint,done,training_iteration,trial_id,date,time_this_iter_s,time_total_s,pid,hostname,node_ip,time_since_restore,iterations_since_restore,config/train_loop_config/depth,config/train_loop_config/hidden_size,config/train_loop_config/ffn_num_layers,config/train_loop_config/ffn_hidden_size,config/train_loop_config/dropout,logdir
0.7342499429068994,1.2729947566986084,1.2729947566986084,0,7,1712674096,checkpoint_000000,True,True,1,c245da43,2024-04-09_10-48-16,10.62180209159851,10.62180209159851,1338742,estes,10.114.0.73,10.62180209159851,1,2,800.0,4,1500.0,0.05,c245da43
0.5953731462941505,1.2962969541549683,1.2962969541549683,0,7,1712674099,checkpoint_000000,True,True,1,0e304808,2024-04-09_10-48-19,10.152503967285156,10.152503967285156,1338851,estes,10.114.0.73,10.152503967285156,1,4,500.0,4,2000.0,0.0,0e304808
```